Content.Server/Advertise/AdvertiseSystem.cs

```diff
@@ -32,102 +35,111 @@ public override void Initialize()
-                _chat.TrySendInGameICMessage(advertise.Owner, Loc.GetString(_random.Pick(advertisements.Advertisements)), InGameICChatType.Speak, true);
-
-            if(refresh)
-                RefreshTimer(uid, true, advertise);
+                _chat.TrySendInGameICMessage(uid, Loc.GetString(_random.Pick(advertisements.Advertisements)), InGameICChatType.Speak, true);
         }
 
-        public void SetEnabled(EntityUid uid, bool enabled, AdvertiseComponent? advertise = null)
+        public void SetEnabled(EntityUid uid, bool enable, AdvertiseComponent? advertise = null)
         {
             if (!Resolve(uid, ref advertise))
                 return;
 
-            var attemptEvent = new AdvertiseEnableChangeAttemptEvent(enabled, advertise.Enabled);
-            RaiseLocalEvent(uid, attemptEvent, false);
+            if (advertise.Enabled == enable)
+                return;
+
+            var attemptEvent = new AdvertiseEnableChangeAttemptEvent(enable);
+            RaiseLocalEvent(uid, attemptEvent);
 
             if (attemptEvent.Cancelled)
                 return;
 
-            if(enabled)
-                RefreshTimer(uid, !advertise.Enabled, advertise);
-
-            advertise.Enabled = enabled;
+            advertise.Enabled = enable;
+            RefreshTimer(uid, advertise);
         }
 
-        private void OnPowerReceiverEnableChangeAttempt(EntityUid uid, ApcPowerReceiverComponent component, AdvertiseEnableChangeAttemptEvent args)
+        private static void OnPowerReceiverEnableChangeAttempt(EntityUid uid, ApcPowerReceiverComponent component, AdvertiseEnableChangeAttemptEvent args)
         {
-            if(args.NewState && !component.Powered)
+            if(args.Enabling && !component.Powered)
                 args.Cancel();
         }
 
-        private void OnVendingEnableChangeAttempt(EntityUid uid, VendingMachineComponent component, AdvertiseEnableChangeAttemptEvent args)
+        private static void OnVendingEnableChangeAttempt(EntityUid uid, VendingMachineComponent component, AdvertiseEnableChangeAttemptEvent args)
         {
-            // TODO: Improve this...
-            if(args.NewState && component.Broken)
+            if(args.Enabling && component.Broken)
                 args.Cancel();
         }
 
         public override void Update(float frameTime)
         {
-            _timer += frameTime;
-
-            if (_timer < UpdateTimer)
-                return;
-
-            _timer -= UpdateTimer;
-
             var curTime = _gameTiming.CurTime;
+            if (_nextCheckTime > curTime)
+                return;
 
-            foreach (var advertise in EntityManager.EntityQuery<AdvertiseComponent>())
+            _nextCheckTime = curTime + _maximumNextCheckDuration;
+            foreach (var (transform, advertise) in EntityManager.EntityQuery<TransformComponent, AdvertiseComponent>())
```

Comment:
Better use an EntityQueryEnumerator pattern, also query the advertise component first since its WAY less common than the transform component, thus the query will run through way less entities when searching
Also remember we have proxy methods